Source/NSubstitute/Core/CallCollection.cs

```diff
@@ -0,0 +1,56 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NSubstitute.Core
+{
+    public class CallCollection : ICallCollection, IReceivedCalls
+    {
+        ConcurrentQueue<CallWrapper> _callWrappers = new ConcurrentQueue<CallWrapper>();
+
+        public void Add(ICall call)
+        {
+            if (call == null) throw new ArgumentNullException(nameof(call));
+            _callWrappers.Enqueue(new CallWrapper(call));
+        }
+
+        public void Delete(ICall call)
+        {
+            if (call == null) throw new ArgumentNullException(nameof(call));
+
+            var callWrapper = _callWrappers.FirstOrDefault(w => w.Call.Equals(call));
+            if (callWrapper == null) throw new InvalidOperationException("Collection doesn't contain the call.");
+
+            callWrapper.IsDeleted = true;
```

Comment:
I've changed implementation so now [it doesn't leak the memory more](https://github.com/Zvirja/NSubstitute/blob/thread-local-last-call-info/Source/NSubstitute/Core/CallCollection.cs#L58).

However, [this test](https://github.com/Zvirja/NSubstitute/blob/thread-local-last-call-info/Source/NSubstitute.Acceptance.Specs/PerfTests.cs#L60) still fails because there is another source of memory leak:
```
->  00000004005d64c8 NSubstitute.Core.CallResults
->  00000004005d64e8 System.Collections.Concurrent.ConcurrentQueue`1[[NSubstitute.Core.CallResults+ResultForCallSpec, NSubstitute]]
->  00000004005d6518 System.Collections.Concurrent.ConcurrentQueue`1+Segment[[NSubstitute.Core.CallResults+ResultForCallSpec, NSubstitute]]
->  000000027fff36c8 System.Collections.Concurrent.ConcurrentQueue`1+Segment[[NSubstitute.Core.CallResults+ResultForCallSpec, NSubstitute]]
->  000000027fff37b8 NSubstitute.Core.CallResults+ResultForCallSpec[]
->  000000027fffc550 NSubstitute.Core.CallResults+ResultForCallSpec
->  000000027fffc238 NSubstitute.Core.ReturnValue
->  0000000576801020 System.Byte[]
```

Therefore, in fact, my change in implementation didn't make any efficient change:
- If call was not configured, its arguments were rooted by `CallStack`.
- If call was configured, its arguments were rooted by `CallResults`. Additionally, they started to be rooted by `CallCollection`, but that shouldn't change the picture.

Nevertheless, now that is fixed, so it's time for another test.